motionplan/planManager.go

```diff
@@ -107,244 +108,294 @@ func (pm *planManager) PlanSingleWaypoint(ctx context.Context,
-								pm.logger.Debugf("fallback path with score %f worse than original score %f; using original", altCost, score)
-							}
-						}
+			}
+
+			// If there was no error, check path quality. If sufficiently good, move on.
+			// If there *was* an error, then either the fallback will not error and will replace it, or the error will be returned
+			if finalSteps.err() == nil {
+				if fallbackPlanner != nil {
+					if ok, score := goodPlan(finalSteps, pathPlanner.opt()); ok {
+						pm.logger.Debugf("got path with score %f, close enough to optimal %f", score, maps.optNode.cost)
+						fallbackPlanner = nil
+					} else {
+						pm.logger.Debugf("path with score %f not close enough to optimal %f, falling back", score, maps.optNode.cost)
+
+						// If we have a connected but bad path, we recreate new IK solutions and start from scratch
+						// rather than seeding with a completed, known-bad tree
+						mapSeed = nil
 					}
 				}
-				// If the fallback wasn't done, we need to get the smoothed path out of the channel
-				if !smoothingDone {
-					finalSteps.steps = <-smoothChan
-				}
+			}
+
+			// Start smoothing before initializing the fallback plan. This allows both to run simultaneously.
+			smoothChan := make(chan []node, 1)
+			utils.PanicCapturingGo(func() {
+				smoothChan <- pathPlanner.smoothPath(ctx, finalSteps.steps)
+			})
+			var alternateFuture *resultPromise
 
-				if finalSteps.err() != nil {
-					return nil, finalSteps.err()
+			// Run fallback only if we don't have a very good path
+			if fallbackPlanner != nil {
+				_, alternateFuture, err = pm.planSingleAtomicWaypoint(
+					ctx,
+					goal,
+					seed,
+					fallbackPlanner,
+					mapSeed,
+				)
+				if err != nil {
+					alternateFuture = nil
 				}
+			}
 
-				if iter+1 < len(goals) {
-					// in this case, we create the next step (and thus the remaining steps) and the
-					// step from our iteration hangs out in the buffer until we're done with it
-					remainingSteps, err = pm.planMotion(
-						ctx,
-						goals,
-						finalSteps.steps[len(finalSteps.steps)-1].Q(),
-						opts,
-						nil,
-						iter+1,
-					)
-					if err != nil {
-						return nil, err
+			// Receive the newly smoothed path from our original solve, and score it
+			finalSteps.steps = <-smoothChan
+			_, score := goodPlan(finalSteps, pathPlanner.opt())
+
+			// If we ran a fallback, retrieve the result and compare to the smoothed path
+			if alternateFuture != nil {
+				alternate, err := alternateFuture.result(ctx)
+				if err != nil {
+					// If the fallback successfully found a path, check if it is better than our smoothed previous path.
+					// The fallback should emerge pre-smoothed, so that should be a non-issue
+					altCost := EvaluatePlan(alternate, pathPlanner.opt().DistanceFunc)
+					if altCost < score {
+						pm.logger.Debugf("replacing path with score %f with better score %f", score, altCost)
+						finalSteps = &rrtPlanReturn{steps: stepsToNodes(alternate)}
```

Comment:
I notice that we never smooth the fallback plan if its better than the smooth path from the original solve, since the smoothing operation is very fast, is there any reason (other than performance) we wouldn't want to apply the smoothing operation prior to returning the fallback as well?

Reply:
We also didn't do this before, but it struck me in both places.

Reply:
The fallback plan is returned pre-smoothed, as when it is produced, it goes through this same code as the non-fallback plan. So there is no need to smooth it a second time.